include/onnxruntime/core/framework/allocator.h

```diff
@@ -15,21 +15,78 @@
 #include "core/framework/fence.h"
 #include "core/session/onnxruntime_c_api.h"
 
+// Struct to represent a physical device.
+struct OrtDevice {
+  typedef int8_t DeviceType;
+  typedef int8_t MemoryType;
+  typedef int16_t DeviceId;
+
+  // Pre-defined device types.
+  static const DeviceType CPU = 0;
+  static const DeviceType GPU = 1; //CUDA
+  static const DeviceType FPGA = 2;
+
+  // Pre-defined memory types.
+  static const MemoryType Default = 0;
+  static const MemoryType CUDA_PINNED = 1;
+
+  constexpr OrtDevice(DeviceType device_type_, MemoryType memory_type_, DeviceId device_id_)
+      : device_type(device_type_),
+        memory_type(memory_type_),
+        device_id(device_id_) {}
+
+  constexpr OrtDevice() : OrtDevice(CPU, Default, 0) {}
+
+  DeviceType Type() const {
+    return device_type;
+  }
+
+  MemoryType MemType() const {
+    return memory_type;
+  }
+
+  DeviceId Id() const {
+    return device_id;
+  }
+
+  inline std::string ToString() const {
+    std::ostringstream ostr;
```

Comment:
Is it worth calculating this value in the ctor and saving it once, or is this never called in a performance sensitive place?

side note: 'inline' isn't necessary when the definition is within the declaration of the class/struct.

Reply:
make sense.

Reply:
just realize that the constructor is asking for "constexpr", which will be done in compilation time, so building them in ctor is not applicable.
Yes, this func is never called in a performance sensitive place. 
"inline" is removed. Thanks!